runtime/compiler/optimizer/InlinerTempForJ9.cpp

```diff
@@ -866,80 +831,155 @@ TR_J9InlinerPolicy::createAnchorNodesForUnsafeGetPut(TR::TreeTop* treeTop,
    {
    TR::CFG *cfg = comp()->getFlowGraph();
-   TR::Block *nullComparisonBlock = prevTreeTop->getEnclosingBlock();
-   TR::TreeTop* nullComparisonTree = nullComparisonBlock->getLastRealTreeTop();
-   TR::TreeTop *nullComparisonEntryTree = nullComparisonBlock->getEntry();
-   TR::TreeTop *nullComparisonExitTree = nullComparisonBlock->getExit();
-   //if conversionNeeded is true, we haven't generated and we don't need arrayDirectAccessBlock
-   TR::Block *arrayDirectAccessBlock = conversionNeeded ? nullComparisonTree->getNode()->getBranchDestination()->getNode()->getBlock() : NULL;
+
+   // Generate the tree for the null comparison
+   TR::Node *addrLoad =
+      TR::Node::createWithSymRef(unsafeAddress,
+                      comp()->il.opCodeForDirectLoad(unsafeAddress->getDataType()),
+                      0, newSymbolReferenceForAddress);
+   TR::Node *nullComparisonNode =
+      TR::Node::createif(TR::ifacmpeq, addrLoad,
+                        TR::Node::create(addrLoad, TR::aconst, 0, 0), NULL);
+   TR::TreeTop *nullComparisonTree = TR::TreeTop::create(comp(), nullComparisonNode);
+
+   // If low-order bit of offset is set, perform indirect access.  The IL for the case
+   // that handles conversion places the indirect access code on the fall-through
+   // path, so it needs to branch if the bit is not set; otherise, the indirect
+   // access code is on the taken path of the branch, so it needs to branch if the
+   // bit is set.
+   //
+   TR::TreeTop *lowTagCmpTree = genClassCheckForUnsafeGetPut(unsafeOffset, /* branchIfLowTagged */ !conversionNeeded);
+
+   TR::TreeTop *firstComparisonTree;
+   TR::TreeTop *ifTree;
+   TR::TreeTop *elseTree;
+
+   TR_OpaqueClassBlock *javaLangClass = comp()->getClassClassPointer(/* isVettedForAOT = */ true);
+
+   // Determine overall layout of IL - which test to place first (null test of object or
+   // setting of low-order bit of offset) - and whether direct or indirect access IL will
+   // be placed on the fall-through path of the first branch.
+   //
+   if (conversionNeeded)
+      {
+      firstComparisonTree = nullComparisonTree;
+      ifTree = arrayDirectAccessTreeTop;
+      elseTree = indirectAccessTreeTop;
+      }
+   else
+      {
+      firstComparisonTree = (javaLangClass != NULL) ? lowTagCmpTree : nullComparisonTree;
+      ifTree = indirectAccessTreeTop;
+      elseTree = directAccessTreeTop;
+      }
+
+   TR::Block *joinBlock =
+         createUnsafeGetPutCallDiamond(
+               callNodeTreeTop, firstComparisonTree, ifTree, elseTree);
+
+   debugTrace(tracer(), "\t In genCodeForUnsafeGetPut, joinBlock is %d\n", joinBlock->getNumber());
+
+   TR::Block *beforeCallBlock = prevTreeTop->getEnclosingBlock();
+
+   TR::Block *arrayDirectAccessBlock = NULL;
    TR::Block *indirectAccessBlock;
-   TR::Block * directAccessBlock;
+   TR::Block *directAccessBlock;
+
    if (conversionNeeded)
       {
+      indirectAccessBlock = beforeCallBlock->getNextBlock();
```

Comment:
I assume "indirect access" here refers to the case when we go through the ram statics on the J9Class. Perhaps we should use similar terminology to what we did in the above code that dealt with treetops (whereas this one deals with blocks) in that we should refer to the direct access as "array direct access block"

Reply:
For the case where conversion is needed, we have a block that performs "direct access" without conversion (for a load or store that is determined to be accessing an instance field), "array direct access" with conversion (for a load or store that is determined to be accessing an array element or native memory) and "indirect access" accessing ram statics for a static field.  So we need to refer to all three of "indirect access", "direct access" and "array direct access" here.

Do you have any thoughts about better terminology for the three kinds of access that might make this easier to understand?  Perhaps "static field access", "full-width access", "minimum-width access"?  Or I could more clearly define the three existing terms, and continue to use those terms.

Reply:
I think it is fine to leave the terminology as you have here. My confusion was because of `ifTree` and `elseTree` above meaning the opposite of what I thought when I added this comment. With the earlier clarification for what that meant, I think the naming between trees and blocks variables in this part of the code is consistent.